### ISSUE TITLE
C++ using directives are not understood (Origin: bugzilla #617285)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#3760]
status NEW severity *normal* in component *general* for *---*
Reported in version *1.6.1* on platform *Other*
Assigned to: Dimitri van Heesch

On 2010-04-30 13:39:06 +0000, Yann Leprince wrote:

    C++ using directives are not understood. For example in the following code,
    the  generated documentation contains an entry for A::var but no entry for
    B::var. Adding "/// \brief The using directive" right before the using
    directive does not change the output.
    
    This behaviour contradicts the meaning of this C++ code: the variable is
    made available in both namespaces.
    
    My feeling is that an entry for var should be generated in the documentation
    of namespace B, containing a pointer to the documentation of var in
    namespace A. In addition, if a documentation block corresponds to the using
    directive, its contents should appear in the documentation of B::var as well
    as the pointer to the documentation of A::var.
    
    
    /// \brief Namespace containing a real variable
    namespace A {
      /// \brief The variable
      int var;
    }
    
    /// \brief Namespace containing using directive
    namespace B {
      using A::nb;
    }

**Comment by nanowizard** [comment]
Having Doxygen handle using directives in this manner is also valuable for class declarations which provide base class member functions via the `using` directive, as outlined in [this SO question](https://stackoverflow.com/questions/36421785/doxygen-c-document-privately-inherited-members-exposed-by-the-using-declar/54887733)

**Comment by robch** [comment]
I would really love to have this issue fixed, because without it I'm going to have to do very unnatural things to keep using `doxygen` in our project.

**Comment by albert-github** [comment]
@robch 
- Which version of doxygen did you try? (As there might already be some improvements in the newer versions)
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

**Comment by eramongodb** [comment]
> This issue should probably be renamed.

This effects all types of `using` that import declarations of entities from one namespace into another.

> Which version of doxygen did you try? (As there might already be some improvements in the newer versions)

There have been no notable improvements since the comment above, and the (fixed) example given in the original 2010 report still fails exactly as described:

```cpp
/// \brief Namespace containing a real variable
namespace A {
  /// \brief The variable
  int var;
}

/// \brief Namespace containing using directive
namespace B {
  using A::var; // <-- This is completely missing in Doxygen results.
}
```

This issue is still present in the latest Doxygen release (1.9.8) and is reproducible by running `doxygen -g && doxygen` in a directory containing just a header file with the contents above.

> There is probably not much value in having Doxygen understand those.

As shown in the example below, it can directly impact the quality of generated documentation in any circumstance where the imported declarations are intended to be part of the documented interface (which is presumably why the using-declaration or using-directive is present to begin with).

> Can you please attach a, small, self contained example

Here is a more detailed example comparing the generated documentation for two headers that both simulate ABI symbol versioning:

```cpp
/// \file ok.hpp Demonstrates desired Doxygen behavior.

/// Generates the desired Doxygen results.
namespace with_inline {
  /// ABI version 1.
  inline namespace v1 {
    /// This is the current `with_inline::S`.
    struct S {};
  }

  /// ABI version 2.
  namespace v2 {
    /// This is an upcoming `with_inline::S`.
    struct S {};
  }
}
```

```cpp
/// \file not_ok.hpp Demonstrates the using-declaration issue.

/// Generates undesirable Doxygen results.
namespace with_using {
  /// ABI version 1.
  namespace v1 {
    /// This is the current `with_using::S`.
    struct S {};
  }

  /// ABI version 2.
  namespace v2 {
    /// This is an upcoming `with_using::S`.
    struct S {};
  }

  using v1::S; // <-- This is completely missing in Doxygen results.
}
```

As before, the issue is reproducible with the latest release Doxygen release (1.9.8) by running `doxygen -g && doxygen` in a directory containing just these two header files.

The `ok.hpp` header generates a `namespacewith__inline.html` page that looks as follows:

> Generates the desired Doxygen results.
> 
> ## Namespaces
> * namespace `v1`: ABI version 1
> * namespace `v2`: ABI version 2
> 
> ## Classes
> * struct `S`: This is the current `with_inline::S`.
> 
> ## Detailed Description
> Generates the desired Doxygen results.

The `not_ok.hpp` header generates a `namespacewith__using.html` page that looks as follows:

> Generates undesirable Doxygen results.
>
> ## Namespaces
> * namespace `v1`: ABI version 1
> * namespace `v2`: ABI version 2
>
> ## Detailed Description
> Generates undesirable Doxygen results.

The `with_using` namespace page should have an entry mentioning "the current `with_using::S`" that redirects to the `with_using::v1::S` class page, just like how the `with_inline` namespace page has an entry that redirects to the `with_inline::vA::S` class page.

Similarly, the generated `annotated.html` page that lists documented classes looks as follows:

> Here are the classes, structs, unions and interfaces with brief descriptions:
>
> * N: `with_inline`: Generates the desired Doxygen results
>   * N: `v1`: ABI version 1.
>     * C: `S`: This is the current `with_inline::S`.
>   * N: `v2`: ABI version 2.
>     * C: `S`: This is an upcoming `with_inline::S`.
>   * C: `S`: This is the current `with_inline::S`.
> * N: `with_using`: Generates undesirable Doxygen results.
>   * N: `v1`: ABI version 1.
>     * C: `S`: This is the current `with_using::S`.
>   * N: `v2`: ABI version 2.
>     * C: `S`: This is an upcoming `with_using::S`.

The `with_using` namespace should have an entry mentioning "the current `with_using::S`" that redirects to the `with_using::v1::S` class page, just like how `with_inline` namespace has an entry that redirects to the `with_inline::v1::S` class page.

Note, attempts to manually document the using-declaration also have no effect:

```cpp
/// This should be documented!
using v1::S; // <-- This is completely missing in Doxygen results.
```

**Comment by bisht2050** [comment]
Hi @albert-github, any updates on this? Could we please fix this sooner than later?

**Comment by doxygen** [comment]
@bisht2050 I'll look into this. I agree this should be handled better.

**Comment by doxygen** [comment]
@bisht2050 Please verify if the referenced commit fixes the problem for you. Do not close the issue, this will be done automatically when the next official release becomes available.

**Comment by eramongodb** [comment]
Thank you very much for the update.

Testing Doxygen with the reference commit (`doxygen -v -> 1.13.0 (d067a7ffe6b2abe94595268487ca3461a62da0a5)`) produces promising results. The cases described earlier in this issue now generate the expected output. However, there are still unresolved cases not yet being handled properly by Doxygen even with the referenced commit.

Each of the snippets below are standalone header files which may be used to reproduce the described issue (labeled with "Problem:") using `doxygen -g && doxygen`. Accompanying Doxygen warnings in command output are included.

---

```cpp
/// @file
/// Problem: @ref B::foo is documented but not a valid reference.

/// Declares @ref A::foo.
namespace A {

/// Declared in @ref A.
void foo();

} // namespace A

/// Expecting B::foo.
namespace B {

using A::foo; // OK: documented in namespace page.

} // namespace B
```

```
header.hpp:2: warning: unable to resolve reference to 'B::foo' for \ref command
```

---

```cpp
/// @file
/// Problem: @ref Y::foo is not documented but is a valid reference.

/// No parent namespace.
struct A {
    /// Member declared in A.
    void foo();
};

/// A parent namespace.
namespace N {

/// Declared in namespace @ref N.
struct B {
    /// Member declared in @ref N::B.
    void foo();
};

} // namespace N

/// Expecting @ref X::foo.
struct X : private A {
    using A::foo; // OK: "Expecting X::foo"
};

/// Expecting @ref Y::foo.
struct Y : private N::B {
    using N::B::foo; // Problem: missing in class page.
};
```

```
<no warnings or errors>
```

---

```cpp
/// @file
/// Problem: @ref B::Y is not documented or a valid reference.

/// Declares X and Y.
namespace A {

/// X declared in A.
class X {};

/// Y declared in A.
class Y {};

} // namespace A

/// Expecting members X and Y.
namespace B {

using   A::X; // OK: "X declared in A."
using ::A::Y; // Problem: missing in namespace page.

} // namespace B
```

```
header.hpp:2: warning: unable to resolve reference to 'B::Y' for \ref command
```

**Comment by doxygen** [comment]
@eramongodb Thanks for the quick feedback. The latest commit should address the issues you found. Let me know if you still see problems.

**Comment by albert-github** [comment]
@doxygen 
I'm a bit confused by the change in scanner.l:
```
- <FindMembers>"using"{BN}+               {
+ <FindMembers>"using"{BN}+/("::"{ID}("::"{ID}*))? {
```
especially the placement of the `*`, shouldn't it be after the first closen `)` ? 
Also isn't `("::"{ID})?` sufficient?

**Comment by doxygen** [comment]
@albert-github Regex is indeed wrong, well spotted! Should be fixed with [this](https://github.com/doxygen/doxygen/commit/c2c9dc36c9b8054ee51fd84457a9f7c3beff5846) commit. The rule is needed to select the using rule for e.g. `using::A::B::C` where `using` is a statement over `Using::A::B::C` where `Using` is a class/namespace.

**Comment by albert-github** [comment]
That the rule is necessary I don't doubt but what I meant is, wouldn't:
```
<FindMembers>"using"{BN}+("::")?/({ID}("::"{ID})*)? 
```
be better as it would remove the first "::" (the "global") part, especially as we see now that the scanner debug output gives (line numbers are 1 off):
**without my proposed change**:
```
--accepting rule at line 1972 ("using   ")
--accepting rule at line 1980 ("A::X")
--accepting rule at line 2042 (";")
--accepting rule at line 953 (" ")
--accepting rule at line 7398 ("//")
--accepting rule at line 7324 (" OK: "X declared in A."
")
--accepting rule at line 953 ("
")
--accepting rule at line 1972 ("using ")
--accepting rule at line 7392 (":")
--accepting rule at line 7392 (":")
--accepting rule at line 1980 ("A::Y")
--accepting rule at line 2042 (";")
--accepting rule at line 953 (" ")
--accepting rule at line 7398 ("//")
--accepting rule at line 7324 (" Problem: missing in namespace page.
")
```
where line 7392 is the "bit bucket"

with  **my proposed change**
```
--accepting rule at line 1972 ("using   ")
--accepting rule at line 1980 ("A::X")
--accepting rule at line 2042 (";")
--accepting rule at line 953 (" ")
--accepting rule at line 7398 ("//")
--accepting rule at line 7324 (" OK: "X declared in A."
")
--accepting rule at line 953 ("
")
--accepting rule at line 1972 ("using ::")
--accepting rule at line 1980 ("A::Y")
--accepting rule at line 2042 (";")
--accepting rule at line 953 (" ")
--accepting rule at line 7398 ("//")
--accepting rule at line 7324 (" Problem: missing in namespace page.
")
```

The explanation:
> The rule is needed to select the using rule for e.g. using ::A::B::C where using is a statement over Using ::A::B::C where Using is a class/namespace.

I don't understand as "using" and "Using" are different words, the scanner.l is case sensitive so "Using" does not get into this rule.

**Comment by doxygen** [comment]
> I don't understand as "using" and "Using" are different words, the scanner.l is case sensitive so "Using" does not get into this rule.

I meant that without the extension of the rule `using ::A::B` would be matched by the rule `<FindMembers,FindMemberName>{SCOPENAME}` because `using` is not seen as a keyword but as part of the name. The leading `::` is now stripped as well. Probably it would be better to pass this down, as the name of the using declaration, so a global scope lookup can be enforced.

**Comment by albert-github** [comment]
> I meant that without the extension of the rule using ::A::B would be matched by the rule 

That is correct.
So probably my proposed change in https://github.com/doxygen/doxygen/issues/3760#issuecomment-2308757896 would be sufficient.

**Comment by doxygen** [comment]
The last commit makes doxygen select the correct class to import in case a leading `::` is used. Example:

```c++
// \file
/// Problem: \ref B::Y is not documented or a valid reference.

/// Declares X and Y.
namespace A {

/// X declared in A.
class X {};

/// Y declared in A.
class Y {};

} // namespace A

/// Expecting members X and Y.
namespace B {

  /// nested A
  namespace A {
    /// Y declared in B::A::Y
    class Y{};
  }

using   A::X; // OK: "X declared in A."
using ::A::Y; // Force to pick A::Y, not B::A::Y
//using   A::Y; // Picks B::A::Y over A::Y

} // namespace B
```

**Comment by albert-github** [comment]
@doxygen commit [9023781](https://github.com/doxygen/doxygen/commit/9023781b60817c46dec9d640b99ce9393cf46a59)is probably better as it solves the underlying problem whilst before and with my fix just the initial `::` was removed.

**Comment by eramongodb** [comment]
Confirmed https://github.com/doxygen/doxygen/commit/9023781b60817c46dec9d640b99ce9393cf46a59 addresses the issues listed prior.

Some other issues discovered while testing with https://github.com/doxygen/doxygen/commit/9023781b60817c46dec9d640b99ce9393cf46a59 (AFAICT these are pre-existing, not regressions):

---

The pages of classes declared via using-declaration do not contain any links or redirects to the original class declaration. Is this by design? (The conditions that lead to a using-declaration generating its own doc page rather than a redirect to the original declaration isn't clear to me yet. `INHERIT_DOCS` and `HIDE_UNDOC_*` might also be related...?)

```cpp
/// \file

/// Declares \ref A::S.
namespace A {

/// Declared in \ref A.
struct S {};

} // namespace A

/// Declares \ref B::S.
namespace B {

// Problem?: \ref B::S does not redirect to \ref A::S.
// Problem?: class page for \ref B::S does not contain any links to \ref A::S.
using A::S;

} // namespace B
```
```
<no warnings or errors>
```

---

Using-declarations of member _operators_ (as opposed to named member functions) are not handled properly:

```cpp
/// \file
/// Problem: doc for \ref B::operator[](int) is missing (but the reference is valid and redirects to \ref A::operator[](int)).
/// Problem: doc for \ref B::operator int() is missing (but the reference is valid and redirects to \ref A::operator int()).

/// Declares A.
struct A {
  /// Declares A::foo().
  void foo();

  /// Declares A::operator[](int).
  void operator[](int);

  /// Declares A::operator int().
  operator int();
};

/// Declares B.
struct B : private A {
  using A::foo;          // OK: documented.
  using A::operator[];   // Problem: missing in class page.
  using A::operator int; // Problem: missing in class page.
};
```
```
<no warnings or errors>
```

---

Using-declarations of forward-declared classes are not handled properly:

```cpp
/// \file
/// Problem: \ref C::S does not exist.

/// Declares namespace A.
namespace A {

struct S;

} // namespace A

/// Declares namespace B.
namespace B {

using A::S; // Problem: C::S is completely missing.

} // namespace B

namespace A {

/// Declares class A::S.
struct S {
  /// Declares member function S::foo.
  void foo();
};

} // namespace A
```
```
header.hpp:1: warning: unable to resolve reference to 'C::S' for \ref command
```

---

Similarly, "transitive"(?) using-declarations within forward-declared classes are not handled properly:

```cpp
/// \file
/// Problem: \ref C::S::foo is missing in class page of \ref C::S (but the
/// reference is valid and redirects to \ref A::S::foo).

/// Declares namespace A.
namespace A {

struct S; // Forward declaration.

} // namespace A

/// Declares namespace B.
namespace B {

struct S; // Forward declaration.

} // namespace B

/// Declares namespace C.
namespace C {

using B::S; // OK: class page for C::S is generated, but...
            // Problem: member function foo is missing in docs for C::S.

} // namespace C

namespace A {

/// Declares class A::S.
struct S {
  /// Declares member function S::foo.
  void foo();
};

} // namespace A

namespace B {

/// Declares class B::S.
struct S : private A::S {
  using A::S::foo; // OK: present in class page.
};

} // namespace B
```
```
<no warnings or errors>
```

**Comment by albert-github** [comment]
@eramongodb Could you check again as so changes were mad in version 34eeb0a5fa77c858b21b47a7533ebe6c46179123

**Comment by eramongodb** [comment]
@albert-github Testing with https://github.com/doxygen/doxygen/commit/34eeb0a5fa77c858b21b47a7533ebe6c46179123 does not appear to address the issues described in https://github.com/doxygen/doxygen/issues/3760#issuecomment-2310744813. Docs for using-declarations of member operators are still missing, and the using-declaration of a forward-declared class is still missing docs of using-declared members.

**Comment by albert-github** [comment]
@eramongodb 

There is a commit made that solves issue #11111 (i.e.commit [2b8fe35](https://github.com/doxygen/doxygen/commit/2b8fe350ec6ea43ceca8d4b9d00afc4c15b1820b)) can you please check whether this solves also the issues mentioned by you in this issue.

**Comment by eramongodb** [comment]
@albert-github Comparing the output of https://github.com/doxygen/doxygen/commit/2b8fe350ec6ea43ceca8d4b9d00afc4c15b1820b relative to https://github.com/doxygen/doxygen/commit/34eeb0a5fa77c858b21b47a7533ebe6c46179123 (excluding `doxygen.css` and `doxygen_crawl.html`):

> Using-declarations of member operators (as opposed to named member functions) are not handled properly: [...]

There is no noticable difference except some removed whitespaces before `</div>` in the source page (`header_8hpp_source.html`):

```diff
- <a href="header_8hpp.html">Go to the documentation of this file.</a>[...]</span> </div>
+ <a href="header_8hpp.html">Go to the documentation of this file.</a>[...]</span></div>
```

> "Transitive"(?) using-declarations within forward-declared classes are not handled properly: [...]

There is no noticable difference except that, in addition to the whitespacing mentioned above, the `B::S` in `using B::S;` is no longer a hyperlink in the source page (this looks like a regression):

```diff
- <div class="line">[...]<span class="keyword">using </span><a class="code hl_struct" href="structB_1_1S.html">B::S</a>; <span class="comment">// OK: class page for C::S is generated, but...</span></div>
+ <div class="line">[...]<span class="keyword">using </span>B::S; <span class="comment">// OK: class page for C::S is generated, but...</span></div>
```

**Comment by doxygen** [comment]
@eramongodb The last commit should address the operator linking, the problems with the transitive example, and the missing link regression.

**Comment by eramongodb** [comment]
@doxygen Confirmed, all outstanding issues mentioned earlier in this thread concerning using-declarations appear to be fixed by https://github.com/doxygen/doxygen/commit/4ebaa8ba334d9421c82b9cb359aa4fb0644f82ee. I do not see any other related issues or regressions at this time. Thank you very much! 👍

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.13.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
